src/msg/async/frames_v2.cc

```diff
@@ -0,0 +1,191 @@
+  for (size_t i = segment_count; i-- > 0; ) {
+    if (segment_bls[i].length() > 0) {
+      return i + 1;
+    }
+  }
+  return 1;
+}
+
+void FrameAssembler::fill_preamble(Tag tag,
+                                   preamble_block_t* preamble) const {
+  // FIPS zeroization audit 20191115: this memset is not security related.
+  ::memset(preamble, 0, sizeof(*preamble));
+
+  preamble->tag = static_cast<__u8>(tag);
+  for (size_t i = 0; i < m_descs.size(); i++) {
+    preamble->segments[i].length = m_descs[i].logical_len;
+    preamble->segments[i].alignment = m_descs[i].align;
+  }
+  preamble->num_segments = m_descs.size();
+  preamble->crc = ceph_crc32c(
+      0, reinterpret_cast<const unsigned char*>(preamble),
+      sizeof(*preamble) - sizeof(preamble->crc));
+}
+
+uint32_t FrameAssembler::get_segment_padded_len(size_t seg_idx) const {
+  return p2roundup<uint32_t>(m_descs[seg_idx].logical_len, CRYPTO_BLOCK_SIZE);
+}
+
+uint32_t FrameAssembler::get_auth_tag_len() const {
+  return m_crypto->rx->get_extra_size_at_final();
+}
+
+// claim_append() doesn't have an rvalue reference overload
+void FrameAssembler::append_encrypt_final(bufferlist* bl) const {
+  auto tmp = m_crypto->tx->authenticated_encrypt_final();
+  bl->claim_append(tmp);
+}
+
+uint64_t FrameAssembler::get_frame_logical_len() const {
+  ceph_assert(!m_descs.empty());
+  uint64_t logical_len = 0;
+  for (size_t i = 0; i < m_descs.size(); i++) {
+    logical_len += m_descs[i].logical_len;
+  }
+  return logical_len;
+}
+
+uint64_t FrameAssembler::get_frame_onwire_len() const {
+  ceph_assert(!m_descs.empty());
+  uint64_t onwire_len = get_preamble_onwire_len();
+  for (size_t i = 0; i < m_descs.size(); i++) {
+    onwire_len += get_segment_onwire_len(i);
+  }
+  onwire_len += get_epilogue_onwire_len();
+  return onwire_len;
+}
+
+void FrameAssembler::asm_crc_r0(const preamble_block_t& preamble,
+                                bufferlist* segment_bls,
+                                bufferlist* frame_bl) const {
+  frame_bl->append(reinterpret_cast<const char*>(&preamble),
+                   sizeof(preamble));
+
+  epilogue_plain_block_t epilogue;
+  // FIPS zeroization audit 20191115: this memset is not security related.
+  ::memset(&epilogue, 0, sizeof(epilogue));
+
+  for (size_t i = 0; i < m_descs.size(); i++) {
+    ceph_assert(segment_bls[i].length() == m_descs[i].logical_len);
+    epilogue.crc_values[i] = segment_bls[i].crc32c(-1);
+    if (segment_bls[i].length() > 0) {
+      frame_bl->claim_append(segment_bls[i]);
+    }
+  }
+  frame_bl->append(reinterpret_cast<const char*>(&epilogue),
+                   sizeof(epilogue));
+}
+
+void FrameAssembler::asm_secure_r0(const preamble_block_t& preamble,
+                                   bufferlist* segment_bls,
+                                   bufferlist* frame_bl) const {
+  bufferlist preamble_bl;
+  preamble_bl.append(reinterpret_cast<const char*>(&preamble),
```

Comment:
Ah, I missed that aspect of `bufferlist`.  Will fix.

Since you've worked a lot on `bufferlist`, a couple of questions:

- When this page is allocated in the first `foo.append()`, the remainder is available for successive `foo.append()` calls, right?

- What about a `foo.append(), foo.claim_append(bar), foo.append()` case?  Does the intervening "claim everything from bar" step inhibit the use of the remainder of the page allocated in the first `append()`?

- Same question for a `foo.append(), bar.splice(&foo), foo.append()` case.  Does the intervening "claim a chunk from from bar" step inhibit the use of the remainder of the page allocated in the first `append()`?

Reply:
Quickly:

1. Right.
2. No, `claim_append()` shouldn't inhibit  and the second `append()` would reuse the space allocated by the first one.
3. No.

Reply:
Thanks!  A couple more places can be improved then (in addition to the ones you've highlighted).